doc: application: add details on example-application

gmarull opened the pull request:
Mention example-application as a reference application.

gmarull committed:
doc: application: add details on example-application

Mention example-application as a reference application.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

nashif commented on doc/application/index.rst:
```diff
@@ -162,10 +162,24 @@ subdirectories which are not described here.
     package.
 
 
+Example Application
```
I think we should make it clear that this example application is based on a manifests, it is not your typical application that we have across the tree without modules. We do not want to have people go through the trouble of setting up something like for a basic application that does not require any of the items listed below.

carlescufi replied:
I agree. Maybe something like "Example standalone application" and then specifically mention west and the [T2 topology](https://docs.zephyrproject.org/latest/guides/west/workspaces.html#topologies-supported) specifically.

mbolivar-nordic reviewed (changes requested):
I think this needs more discussion and infrastructure around the release process.

I am against this change unless we use example-application branches or tags whose manifests will be pinned to the release version of zephyr they were tested on.

So I'm saying for example:

- the zephyr 2.6 docs build should point to an example-application manifest that imports zephyr with `revision: v2.6.0` (or `v2.7.1`, `v2.7.2`, etc.)
- the zephyr 2.7 docs build should point to an example-application manifest that imports zephyr with `revision: v2.7.0`, etc.
- zephyr main can point to an example-application manifest that imports with `revision: main`

gmarull commented:
> I think this needs more discussion and infrastructure around the release process.
> 
> I am against this change unless we use example-application branches or tags whose manifests will be pinned to the release version of zephyr they were tested on.
> 
> So I'm saying for example:
> 
> * the zephyr 2.6 docs build should point to an example-application manifest that imports zephyr with `revision: v2.6.0` (or `v2.7.1`, `v2.7.2`, etc.)
> * the zephyr 2.7 docs build should point to an example-application manifest that imports zephyr with `revision: v2.7.0`, etc.
> * zephyr main can point to an example-application manifest that imports with `revision: main`

I think it should be sufficient to make `example-application` follow Zephyr versioning and mention that in both Zephyr docs and example app README.

gmarull commented:
> Can you explain what that means? Right now we don't have anything except a single branch with a manifest that points to zephyr `main` (in the example-application repository).

I mean, Zephyr docs should just link to https://github.com/zephyrproject-rtos/example-application independently of the Zephyr version. `example-application` should make explicit that it follows Zephyr versioning (I assume the idea is to make `example-application` `v2.6.0` point to Zephyr `v2.6.0`).

mbolivar-nordic commented:
>  `example-application` should make explicit that it follows Zephyr versioning (I assume the idea is to make `example-application` `v2.6.0` point to Zephyr `v2.6.0`).

I think I agree, but "should" is a key word here. Right now it doesn't do anything like that, and I think we should sort that out before merging this.

carlescufi commented on doc/application/index.rst:
```diff
@@ -162,11 +162,26 @@ subdirectories which are not described here.
+
+A reference standalone application can be found in the `Example Application`_
+repository. It can be used as a reference on how to structure Zephyr based
+applications using the :ref:`T2 star topology <west-t2>`. It also demonstrates
+features commonly used in applications such as:
```
```suggestion
applications using the :ref:`T2 star topology <west-t2>`. It also demonstrates the out-of-tree use of features commonly used in applications such as:
```

carlescufi commented on doc/application/index.rst:
```diff
@@ -162,11 +162,26 @@ subdirectories which are not described here.
+
+- Custom boards
+- Custom devicetree bindings
+- Out-of-tree drivers
```
```suggestion
- Custom drivers
```

carlescufi commented on doc/application/index.rst:
```diff
@@ -162,11 +162,26 @@ subdirectories which are not described here.
 
 Follow these steps to create a new application directory. (Refer to
-:ref:`samples-and-demos` for existing applications provided as part of Zephyr.)
+the `Example Application`_ repository for a reference standalone application
```
```suggestion
the `Example Application`_ repository for a reference standalone application in its own Git repository
```

carlescufi commented on doc/application/index.rst:
```diff
@@ -162,11 +162,26 @@ subdirectories which are not described here.
+Example standalone application
+******************************
+
+A reference standalone application can be found in the `Example Application`_
```
```suggestion
A reference standalone application contained in its own Git repository can be found in the `Example Application`_
```

carlescufi commented on doc/application/index.rst:
```diff
@@ -162,11 +162,26 @@ subdirectories which are not described here.
+******************************
+
+A reference standalone application can be found in the `Example Application`_
+repository. It can be used as a reference on how to structure Zephyr based
```
```suggestion
repository. It can be used as a reference on how to structure out-of-tree, Zephyr-based
```

carlescufi reviewed (changes requested):
Much better!

carlescufi commented:
> > `example-application` should make explicit that it follows Zephyr versioning (I assume the idea is to make `example-application` `v2.6.0` point to Zephyr `v2.6.0`).
> 
> I think I agree, but "should" is a key word here. Right now it doesn't do anything like that, and I think we should sort that out before merging this.

Agreed, maybe a quick sentence here or in the example application .md @gmarull?

carlescufi commented:
@gmarull 
As part of this PR, can you also add a line in the "major enhancements" section of the [release notes](https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/releases/release-notes-2.6.rst) with something along the lines of: "Added an example standalone Zephyr application in its own Git repository"

gmarull commented:
> > > `example-application` should make explicit that it follows Zephyr versioning (I assume the idea is to make `example-application` `v2.6.0` point to Zephyr `v2.6.0`).
> > 
> > 
> > I think I agree, but "should" is a key word here. Right now it doesn't do anything like that, and I think we should sort that out before merging this.
> 
> Agreed, maybe a quick sentence here or in the example application .md @gmarull?

Will the example-application follow Zephyr versioning? I understand this has not been formally decided yet (maybe I missed it).

If you guys confirm this will be the rule I can update example-applicacion readme. We should likely include the procedure as part of the release process, too.

gmarull committed:
doc: add example application to the release notes

Mention the example application as a major enhancement and provide a
link to its repository.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull commented:
> Added an example standalone Zephyr application in its own Git repository

done

carlescufi commented:
> > > > `example-application` should make explicit that it follows Zephyr versioning (I assume the idea is to make `example-application` `v2.6.0` point to Zephyr `v2.6.0`).
> > > 
> > > 
> > > I think I agree, but "should" is a key word here. Right now it doesn't do anything like that, and I think we should sort that out before merging this.
> > 
> > 
> > Agreed, maybe a quick sentence here or in the example application .md @gmarull?
> 
> Will the example-application follow Zephyr versioning? I understand this has not been formally decided yet (maybe I missed it).

I believe we did decide that, yes. I was not able to find the relevant comment after a search, but why wouldn't we do this? This is now part of the release, and so it should be tagged all the same.

> If you guys confirm this will be the rule I can update example-applicacion readme. We should likely include the procedure as part of the release process, too.

If this PR makes it to 2.6, then let's update: https://github.com/zephyrproject-rtos/zephyr/issues/32293

gmarull commented:
> I believe we did decide that, yes. I was not able to find the relevant comment after a search, but why wouldn't we do this? This is now part of the release, and so it should be tagged all the same.

I agree we should follow the same versioning, we just need to make sure it's properly documented.

carlescufi commented:
> > I believe we did decide that, yes. I was not able to find the relevant comment after a search, but why wouldn't we do this? This is now part of the release, and so it should be tagged all the same.
> 
> I agree we should follow the same versioning, we just need to make sure it's properly documented.

Let's push a PR to the example-application repo stating that fact then, I do not foresee any controversy.

gmarull commented:
> Let's push a PR to the example-application repo stating that fact then, I do not foresee any controversy.

https://github.com/zephyrproject-rtos/example-application/pull/9
